src/core/thread/mle.cpp

```diff
@@ -4741,5 +4764,56 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    OT_ASSERT(IsChild());
+
+    if (mDetachGracefullyTimer.IsRunning())
+    {
+        ExitNow(error = kErrorBusy);
+    }
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+
+    IgnoreError(SendChildUpdateRequest(/*aTimeout=*/0));
+
+exit:
+    return error;
+}
+
+void Mle::HandleDetachGracefullyTimer(Timer &aTimer)
+{
+    aTimer.Get<Mle>().HandleDetachGracefullyTimer();
+}
+
+void Mle::HandleDetachGracefullyTimer(void)
+{
+    Stop();
+}
+
+void Mle::HandleFinishedDetachingTasklet(Tasklet &aTasklet)
+{
+    aTasklet.Get<Mle>().HandleFinishedDetachingTasklet();
+}
+
+void Mle::HandleFinishedDetachingTasklet(void)
+{
+    if (mDetachGracefullyCallback != nullptr)
```

Comment:
It's possible, even though highly unlikely, that the Thread state may not be disabled at this moment, if any logic executed after posting tasklet but before `HandleFinishedDetachingTasklet`. 

It's a general issue for posting any tasklets.

That being said, I am not sure how to address it elegantly.

Reply:
After posting the tasklet, the state is disabled, so there should be no logic that can change the state running. Otherwise `thread stop` wouldn't be able to guarantee the state either.

Unless the user starts Thread in between. I think that's fine. It doesn't make sense to start Thread before finishing detaching.

Reply:
> After posting the tasklet, the state is disabled, so there should be no logic that can change the state running.

Note that other task lets could be executed before this tasklet. To be conservative, any logic could execute in other tasklets.

Reply:
See https://github.com/openthread/openthread/pull/7666#discussion_r876838646. Now it's impossible to run any logic between Stop() and the callback.